piksi_tools/fileio.py

```diff
@@ -25,39 +24,41 @@
 from piksi_tools import serial_link
 
 MAX_PAYLOAD_SIZE = 255
-SBP_FILEIO_WINDOW_SIZE = 10
-SBP_FILEIO_TIMEOUT = 1.0
+SBP_FILEIO_WINDOW_SIZE = 40
+SBP_FILEIO_TIMEOUT = 5.0
 
 
-class Semaphore(object):
-    """Semaphore implementation with timeout"""
-
-    def __init__(self, n=1):
-        self.sem = threading.Semaphore(n)
-        self.ev = threading.Event()
-
-    def acquire(self, timeout=None):
-        while not self.sem.acquire(False):
-            if not self.ev.wait(timeout):
-                return False
-        self.ev.clear()
-        return True
-
-    def release(self):
-        self.sem.release()
-        self.ev.set()
+class PendingWrite(object):
+    __slots__ = ["message", "sequence", "time", "tries", "complete"]
 
 
 class SelectiveRepeater(object):
     """Selective repeater for SBP file transfers"""
 
     def __init__(self, link, msg_type, cb=None):
-        self.sem = Semaphore(SBP_FILEIO_WINDOW_SIZE)
-        self.window = {}
+        self.window = [PendingWrite() for X in range(SBP_FILEIO_WINDOW_SIZE)]
+        self.pending = []
         self.link = link
         self.msg_type = msg_type
         self.cb = cb
 
+    def _return_pending_write(self, pending_write):
+        rtt = time.time() - pending_write.time
+        self.pending.pop(self.pending.index(pending_write))
+        self.window.append(pending_write)
+
+    def _fetch_pending_write(self, msg):
+        pending_write = self.window.pop()
+        self.pending.append(pending_write)
```

Comment:
i see that you are using the "complete" flag as a guard of sorts, but I think it is more intuitive to pull form queu, set data, then add to pending.

Reply:
Added some methods that make this flow clearer, I'm not sure the `.complete` flag is necessary, but I added for the case when the retry function makes pass through a stale `self.pending` list, it can use the `.complete` flag to know that it doesn't actually need to check that pending write object.